### DISCUSSION TITLE
LLVM /+ MLIR support

### DISCUSSION THREAD

**Discussion by thatTPR** [eclipse-langium/langium#267] [Ideas]
Created from  [issue 259](https://github.com/langium/langium/issues/259) (now closed)

The main thing about this feature I think would be to improve better diagnostic tooling and combine the pattern matching abilities of LSP with the IR tooling of LLVM. This is very speculative at this point since LLVM IR's were not designed for this but if it is proven that compiler technology can be merged with developer tooling in this fashion it should prove a most valuable framework. No more extra errors in the build step ( yay ). At least that is the theory.
I posted this on LSP main as well because at least to me this is more of a grammar support thing and might be better suited there. If it gets more attention from the maintainers I will close this thread with a link there

**Comment by spoenemann** [comment]
Thanks for reaching out on this topic! I have not been much involved with LLVM / MLIR so far, but let's use this thread to gather thoughts. So if anyone is interested, please contribute to the discussion.

**Comment by thatTPR** [comment]
This kinda reduces the MxN problem from an M+N problem to an N problem.
MxN problem is the problem of producing Dev tools for each M editors times each N language. LSP reduces this to an M + N problem. I hope with LLVM/MLIR integration this gets reduced to an M problem. Produce tooling just for the language IR and then perform translation for each language . This might not even  be possible since LLVM/MLIR is a toolchain not a language and does not have a runtime nor is it reasonable to assume that one can even be built. That being said the tooling around this standards is most likely more than enough to build one. This is no easy task . LLVM is probably the biggest single repository ever built and it is all tooling. Thankfully it's all optimization tooling.

**Reply by spoenemann** [reply]
@thatTPR in case you're still interested: combining Langium and LLVM works!
 
👉 https://www.typefox.io/blog/langium-llvm/
